src/main/java/io/gingersnapproject/cdc/cache/hotrod/HotRodService.java

```diff
@@ -40,68 +37,54 @@ public class HotRodService implements CacheService {
 
    @Inject NotificationManager eventing;
 
-   @Inject Configuration config;
-
    @Inject DBSyncerMetrics metrics;
 
-   volatile RemoteCacheManager rcm;
-
-   private final Map<String, CacheBackend> backends = new ConcurrentHashMap<>();
-   private volatile boolean shutdownSignal;
-
-   @PostConstruct
-   public void init() {
-      rcm = new RemoteCacheManager(config.cache().uri());
-      createCaches();
-   }
-
-   public void shutdown(@Observes ShutdownEvent ignore) {
-      if (rcm != null) {
-         shutdownSignal = true;
-         if (backends.values().stream().noneMatch(CacheBackend::isRunning)) destroy();
-      }
-   }
+   private final Map<URI, HotRodCache> managers = new ConcurrentHashMap<>();
 
    @Override
-   public CompletionStage<Boolean> reconnect(String name, Rule rule) {
-      if (!backends.containsKey(name)) return CompletableFutures.completedFalse();
-
-      return CompletableFuture.supplyAsync(() -> {
-         backendForRule(name, rule).start();
-         return null;
-      }).handle((ignore, t) -> {
-         if (t != null) {
-            eventing.backendFailedEvent(name, t);
-            return false;
-         }
-         createCaches();
-         return true;
+   public CompletionStage<Boolean> reconnect(CacheIdentifier identifier, Rule rule) {
+      var cache = getCacheOrThrow(identifier);
+      return cache.rcm.startAsync().thenApply(ignore -> {
+         var backend = cache.backends.get(identifier.rule());
+         if (backend == null) throw new IllegalStateException(String.format("Backend not found for '%s'", identifier));
+         return backend.reconnect();
       });
    }
 
    @Override
-   public void stop(String name) {
-      if (!backends.containsKey(name)) throw new IllegalStateException(String.format("Backend for %s not created", name));
-      backends.computeIfPresent(name, (ignore, cache) -> {
-         cache.stop();
+   public void stop(CacheIdentifier identifier) {
+      managers.computeIfPresent(identifier.uri(), (ignore, cache) -> {
+         var backend = cache.backends.remove(identifier.rule());
+
+         if (backend != null) backend.stop();
+         if (cache.backends.isEmpty()) {
+            cache.rcm.stop();
```

Comment:
```suggestion
            cache.rcm.stopAsync().whenComplete(___, t) -> {
               if (t != null) {
                  // TODO: change this to a nicer message
                  log.error(t);
               }
            });
```
No reason to block in the compute lambda and no requirement to propagate the exception afaik. Let me know if there is a reason to.